### PR TITLE
Changes needed for computing bounds of Shelley header sizes

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

**Comment by coot** [comment]
@dcoutts I updated it against current master.  I also refactored it slightly so it is easy to add new `ToCBOR` classes which would include `Size` expressions. All `KESAlgorithm` and `VRFAlgorithm` instances which have `ToCBOR` instances for their keys / certs / sigs have `encodedSizeExpr` (which proves that the first point works).

File: cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs [hunk 1]
```diff
@@ -70,6 +71,11 @@ instance IsString (Hash h a) where
 
 instance (HashAlgorithm h, Typeable a) => ToCBOR (Hash h a) where
   toCBOR = toCBOR . getHash
+  encodedSizeExpr _size proxy =
+      encodedSizeExpr (\_ -> hashSize) (getHash <$> proxy)
+    where
+      hashSize :: Size
+      hashSize = fromIntegral (sizeHash (Proxy :: Proxy h))
```
Comment (dcoutts) [inline cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs:78]: I don't understand this. It needs explaining.

I presume what we're doing here is reusing the `ByteString` instance (that's what the `(getHash <$> proxy)` gives us) and then passing as the size expression the hash length.

If I'm right, that took me about 10min of studying the `encodedSizeExpr` code to figure out.

File: cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs [hunk 2]
```diff
@@ -261,3 +266,19 @@ encodeSignedKES (SignedKES s) = encodeSigKES s
...
+accessProxy _ = Proxy
+
+encodedVerKeyKESSizeExpr :: KESAlgorithm v => Proxy (VerKeyKES v) -> Size
+encodedVerKeyKESSizeExpr proxy = fromIntegral (sizeVerKeyKES (accessProxy proxy))
+
+encodedSignKeyKESSizeExpr :: KESAlgorithm v => Proxy (SignKeyKES v) -> Size
+encodedSignKeyKESSizeExpr proxy = fromIntegral (sizeSignKeyKES (accessProxy proxy))
+
+encodedSigKESSizeExpr :: KESAlgorithm v => Proxy (SigKES v) -> Size
+encodedSigKESSizeExpr proxy = fromIntegral (sizeSigKES (accessProxy proxy))
```
Comment (dcoutts) [inline cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs:284]: Again, could do with comments. This is clever tricky stuff, with subtle types.

File: cardano-crypto-class/src/Cardano/Crypto/VRF/Class.hs [hunk 3]
```diff
@@ -218,6 +227,17 @@ instance (VRFAlgorithm v, Typeable a) => ToCBOR (CertifiedVRF v a) where
       toCBOR (certifiedNatural cvrf) <>
       encodeCertVRF (certifiedProof cvrf)
 
+  encodedSizeExpr size proxy =
+        1
+      + encodedSizeExpr    size ((toWord64 . certifiedNatural) <$> proxy)
```
Comment (dcoutts) [inline cardano-crypto-class/src/Cardano/Crypto/VRF/Class.hs:232]: This is using a proxy at type Word64 right? But it's actually a natural.

So it really is a `Natural` but the max size of that `Natural` is given in the class above by:
```
  maxVRF :: proxy v -> Natural
```

File: cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs (hunk 1 above)
Comment (coot) [reply cardano-crypto-class/src/Cardano/Crypto/Hash/Class.hs:78]: Yes, it does exactly that :).

**Comment by coot** [comment]
> I'm not sure this kind of style is very clear
>   ```
>   ((toWord64 . certifiedNatural) <$> proxy)
>   ```

Let's take a simple example:
```
data Tuple  a b = Tuple {
    first :: a,
    second :: a
  }

instance (ToCbor a, ToCbor b) => ToCbor (Tuple a b) where
 toCbor tuple = 
                  encodeListLen 2
             <> toCbor (first tuple)
             <> toCbor (sencod tuple)
  encodedSizeExpr size tupleProxy =
                  1
                + encodedSizeExpr size (first <$> tupleProxy)
                + encodedSizeExpr size (second <$> tupleProxy)
```
What I really like about it is that:
1. the two `toCbor` and `encodedSizeExpr` implementation are of the same shape. One is accessing fields using field acessor the other functoriality of `Proxy`. In some cases using  contravariancy also makes sense (I had one such case for `Byron`).
2. if type of `Tuple` changes (specializes), both terms will give a type error.

**Comment by coot** [comment]
> Wouldn't it be clearer to make use of ScopedTypeVariables and say:
> ```
> encodedSigDSIGNSizeExpr :: forall v. DSIGNAlgorithm v => Proxy (SigDSIGN v) -> Size
> encodedSigDSIGNSizeExpr _ = fromIntegral (sizeSigDSIGN (Proxy :: Proxy v))
> ```
I am fine with that, as this keeps connection between the two.

File: cardano-crypto-class/src/Cardano/Crypto/VRF/Class.hs (hunk 3 above)
Comment (coot) [reply cardano-crypto-class/src/Cardano/Crypto/VRF/Class.hs:232]: I see,I misinterpreted it. I will fix this.

File: cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs (hunk 2 above)
Comment (coot) [reply cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs:284]: This is just using `Num` instance.  The reason for this is that `ToCBOR` module does not export `Size` constructors, it forces to either use `Num` instance or `NumDecimals` extension.

File: cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs [hunk 4]
```diff
@@ -261,3 +266,37 @@ encodeSignedKES (SignedKES s) = encodeSigKES s
...
+
+-- | 'Size' expression for 'VerKeyKES' which is using 'sizeVerKeyKES' encoded
+-- as 'Size'.
+--
+encodedVerKeyKESSizeExpr :: forall v. KESAlgorithm v => Proxy (VerKeyKES v) -> Size
+encodedVerKeyKESSizeExpr _proxy =
+      -- 'encodeBytes' envelope
+      fromIntegral ((withWordSize :: Word -> Integer) (sizeVerKeyKES (Proxy :: Proxy v)))
+      -- payload
+    + fromIntegral (sizeVerKeyKES (Proxy :: Proxy v))
```
Comment (dcoutts) [inline cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs:282]: These could also use the ByteString instance with the fixed size, like the Hash instance does.

File: cardano-crypto-class/test/Test/Crypto/Util.hs [hunk 5]
```diff
@@ -96,6 +102,14 @@ prop_cbor :: (ToCBOR a, FromCBOR a, Eq a, Show a)
           => a -> Property
 prop_cbor = prop_cbor_with toCBOR fromCBOR
 
+prop_cbor_size :: forall a. ToCBOR a => a -> Property
+prop_cbor_size a = counterexample (show lo ++ " ≰ " ++ show len) (lo <= len)
+              .&&. counterexample (show len ++ " ≰ " ++ show hi) (len <= hi)
+  where
+    len, lo, hi :: Natural
+    len = fromIntegral $ BS.length (toStrictByteString (toCBOR a))
+    Right (Range {lo, hi}) = szSimplify $ encodedSizeExpr szGreedy (Proxy :: Proxy a)
```
Comment (dcoutts) [inline cardano-crypto-class/test/Test/Crypto/Util.hs:111]: Perfect.

**Comment by coot** [comment]
> I think the key instances could be simpler by following the pattern you used for the hash type of reusing the bytestring instance and passing in the size expr as the fixed length.

One cannot do that for `KES`, `VRF` or `DSIGN` types; they are not just wrapping something that already has `ToCBOR` instance like `Hash` does. `ByteString` instance is clever to use a suspended computation, .

File: cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs (hunk 4 above)
Comment (coot) [reply cardano-crypto-class/src/Cardano/Crypto/KES/Class.hs:282]: Not every instance of `VerKeyKES` is wrapping a `ByteString`. Even though you would get a correct result, I don't think this is right.

**Pull request merged**